Added macro expansion support

smalti opened the pull request:
#### 1. Added new macro: `VERIFY_AND_CLEAR_MODULE_FUNC`

#### 2. Added macro expansion support for:
- RESTORE_MODULE_FUNC
- VERIFY_AND_CLEAR_MODULE_FUNC_EXPECTATIONS
- VERIFY_AND_CLEAR_MODULE_FUNC
- REAL_MODULE_FUNC
- INVOKE_REAL_MODULE_FUNC

#### 3. Renamed `patch_module_func_` to `patch_module_func_non_optimized` and moved it to the private namespace: `gmock_win32::detail`

Alex0vSky commented on include/gmock-win32.h:
Someday we will have to count how many macros we have :).

smalti replied:
> Someday we will have to count how many macros we have :).

I recently thought about how to simplify the client's work with our macroses. Now we have a lot of them, and maybe we need to change the prefix for our implementation-specific macro-defines (to avoid accidental incorrect usage due to іntellisense auto-complete). What are your thoughts on this approach?

Possible prefixes, e.g.: `INTERNAL_*` / `IMPL_*` / `PRIVATE_*`

Alex0vSky reviewed (approved):
OMG, this is about `VERIFY_AND_CLEAR_MODULE_FUNC_EXPECTATIONS`.
When I wrote the first tests, doubts arose about the correctness of its functioning, but I decided to leave all the questions for later.
In short, it was not a priority. I hope there is a fix, but I can’t check right now.

smalti commented:
> OMG, this is about `VERIFY_AND_CLEAR_MODULE_FUNC_EXPECTATIONS`.

I realized I had omitted `VERIFY_AND_CLEAR*` during the initial implementation, so I added it later to align with `GMock`. Now, we have two variants:

```cpp
// Verifies and removes the expectations
VERIFY_AND_CLEAR_MODULE_FUNC_EXPECTATIONS(GetCurrentProcessId);

// Verifies and removes the expectations and removes the default actions set by ON_CALL
VERIFY_AND_CLEAR_MODULE_FUNC(GetProcessIdOfThread);
```

> When I wrote the first tests, doubts arose about the correctness of its functioning, but I decided to leave all the questions for later. In short, it was not a priority. I hope there is a fix, but I can’t check right now.

In `GMock`, the expectations of a mock object are verified upon its destruction. However, since we do not destroy our objects, we must manually verify our expectations after each test. There are different cases when it can have different behavior (perhaps we should mention these nuances in the documentation):

1. In the general case, we should verify expectations after every test case, similar to how a Mock object would be destroyed. This can typically be done in the `TearDown` method.

2. In specific cases, particularly when functions used are also part of `GMock` (e.g. during `::testing::Mock::VerifyAndClear*` calls), we should first call `RESTORE_MODULE_FUNC` and after that, we can call `VERIFY_AND_CLEAR_MODULE_FUNC_*` (so it would work properly with real functions).

smalti commented:
> OMG, this is about `VERIFY_AND_CLEAR_MODULE_FUNC_EXPECTATIONS`.

I'm going to make a new tag `1.2.2` with these changes:

- Added support for Win32 API forwarding
- Added support for macro expansion across all macro definitions
- Added `VERIFY_AND_CLEAR_MODULE_FUNC` macro
- Fixed a typo related to 12 arguments
- Fixed C++20 compilation errors and warning 4702

Alex0vSky commented:
>... we do not destroy our objects ...

Yes, this is it, now everything is clear.
> I'm going to make a new tag 1.2.2

Yeah, a new version is always a holiday! I am glad.
We'll probably introduce "SemVer" soon. It will be possible to connect it with *CI*.
